src/analyzer/vulnerabilities/arbitrary_from_in_transferfrom.rs

```diff
@@ -0,0 +1,124 @@
+use std::collections::HashSet;
+
+use solang_parser::pt::{self, Expression, Loc};
+use solang_parser::{self, pt::SourceUnit};
+
+use crate::analyzer::ast::{self, Target};
+
+pub fn _arbitrary_from_in_transferfrom_vulnerability(source_unit: SourceUnit) -> HashSet<Loc> {
+    //Create a new hashset that stores the location of each vulnerability target identified
+    let mut vulnerability_locations: HashSet<Loc> = HashSet::new();
+
+    //Extract the target nodes from the source_unit
+    let target_nodes =
+        ast::extract_target_from_node(Target::FunctionDefinition, source_unit.into());
+
+    //For each target node that was extracted, check for the vulnerability patterns
+    for _node in target_nodes {
+        let contract_part = _node.contract_part().unwrap();
+
+        if let pt::ContractPart::FunctionDefinition(box_fn_definition) = contract_part {
+            // if function has no params or no body, it is probably not affected by this vulnerability.
+            if box_fn_definition.params.is_empty() || box_fn_definition.body.is_none() {
+                continue;
+            }
+
+            // We extract the function parameters to later verify if they are used as the first parameter in the `transferFrom/SafeTransferFrom`.
+            let fn_params = extract_fn_params_as_string(&box_fn_definition);
+
+            if let pt::Statement::Block {
+                loc: _,
+                unchecked: _,
+                statements,
+            } = box_fn_definition.body.unwrap()
+            {
+                // We loop through each body expression to determine if 'transferFrom/SafeTransferFrom' is used.
+                for _statement in statements {
+                    if let pt::Statement::Expression(_, box_expression) = _statement.clone() {
+                        if let pt::Expression::FunctionCall(
+                            loc,
+                            box_fn_expression,
+                            fn_call_params,
+                        ) = box_expression
+                        {
+                            // If a call exists and any of the function parameters is used as the first parameter in the call, we mark it as a vulnerability.
+                            if should_be_added(box_fn_expression, &fn_params, &fn_call_params[0]) {
```

Comment:
I think we can probably remove the `should_be_added` function and just expose 
```rust
if has_transfer_from_expression(box_fn_expression) && transfer_from_uses_fn_params_as_from_variable(&fn_params, &from_param){
    //add location
}
```

Also if the `from` address is not the first argument, will this catch the vulnerability? I haven't run this locally but at first glance, it seems we would want to check if any passed in address is used as the `from` argument for `transferFrom`.

Reply:
This checks if any of the parameters of the public/external function is used as the first parameter of transferFrom, which, if it follows the standard, should be the first parameter. I don't know if I understood what you were referring to correctly. Likewise, tomorrow I will try if I find any flaw in what I mentioned or, let me know if I misunderstood what you were going to.